File: rag_frontend/src/api/knowledge-graph.ts
import { request } from '@/utils/request'

export interface KnowledgeGraphEntity {
  name: string
  type: string
  properties: Record<string, any>
}

export interface KnowledgeGraphRelation {
  source: string
  target: string
  type: string
  properties: Record<string, any>
}

export interface BuildKnowledgeGraphRequest {
  text: string
  user_id?: string
  session_id?: string
  extract_entities?: boolean
  extract_relations?: boolean
}

export interface BuildKnowledgeGraphResponse {
  success: boolean
  message?: string
  entities: KnowledgeGraphEntity[]
  relations: KnowledgeGraphRelation[]
}

export interface KnowledgeGraphSearchRequest {
  query: string
  user_id?: string
  session_id?: string
  top_k?: number
  vector_weight?: number
  graph_weight?: number
  use_graph?: boolean
}

export interface KnowledgeGraphSearchResult {
  content: string
  score: number
  source: 'vector' | 'graph'
  metadata: Record<string, any>
}

export interface KnowledgeGraphSearchResponse {
  success: boolean
  results: KnowledgeGraphSearchResult[]
  vector_results_count: number
  graph_results_count: number
  total_count: number
}

export interface QueryEntityRequest {
  entity_name: string
  max_depth?: number
  limit?: number
}

export interface QueryEntityResponse {
  success: boolean
  entity: KnowledgeGraphEntity
  relations: Array<{
    relation: string
    target: KnowledgeGraphEntity
    properties: Record<string, any>
  }>
}

export interface GraphVisualizationNode {
  id: string
  label: string
  type: string
  properties: Record<string, any>
}

export interface GraphVisualizationEdge {
  id: string
  source: string
  target: string
  type: string
  properties: Record<string, any>
  description?: string
}

export interface GraphVisualizationResponse {
  nodes: GraphVisualizationNode[]
  edges: GraphVisualizationEdge[]
  center_node: string | null
}

export interface GraphImportRequest {
  nodes: GraphVisualizationNode[]
  edges: GraphVisualizationEdge[]
  deleted_node_ids?: string[]
  deleted_edge_ids?: string[]
}

export interface GraphImportResponse {
  success: boolean
  nodes_saved: number
  edges_saved: number
  nodes_deleted: number
  edges_deleted: number
  errors: string[]
}

export interface VisualizeRequest {
  entity_name?: string
  max_depth?: number
  limit?: number
}

export interface EntityListItem {
  id: string
  name: string
  type: string
  properties: Record<string, any>
  created_at: string | null
  updated_at: string | null
}

export interface EntityListResponse {
  entities: EntityListItem[]
  total: number
  limit: number
  offset: number
}

export interface EntityTypesResponse {
  types: string[]
}

export const knowledgeGraphApi = {
  // 构建知识图谱（增加超时到5分钟，因为LLM提取需要时间）
  async build(data: BuildKnowledgeGraphRequest): Promise<BuildKnowledgeGraphResponse> {
    return request<BuildKnowledgeGraphResponse>('/knowledge_graph/build', {
      method: 'POST',
      data: JSON.stringify(data),
      timeout: 300000,
    })
  },

  // 混合检索
  async search(data: KnowledgeGraphSearchRequest): Promise<KnowledgeGraphSearchResponse> {
    return request<KnowledgeGraphSearchResponse>('/knowledge_graph/search', {
      method: 'POST',
      data: JSON.stringify(data),
    })
  },

  // 查询实体
  async queryEntity(data: QueryEntityRequest): Promise<QueryEntityResponse> {
    return request<QueryEntityResponse>('/knowledge_graph/query-entity', {
      method: 'POST',
      data: JSON.stringify(data),
    })
  },

  // 可视化图谱
  async visualize(data: VisualizeRequest): Promise<GraphVisualizationResponse> {
    const params = new URLSearchParams()
    if (data.entity_name) params.append('entity_name', data.entity_name)
    if (data.max_depth !== undefined) params.append('max_depth', String(data.max_depth))
    if (data.limit !== undefined) params.append('limit', String(data.limit))
    
    return request<GraphVisualizationResponse>(`/knowledge_graph/visualize?${params.toString()}`, {
      method: 'GET',
    })
  },

  // 获取实体列表
  async importGraph(data: GraphImportRequest): Promise<GraphImportResponse> {
    return request<GraphImportResponse>('/knowledge_graph/import', {
      method: 'POST',
      data: JSON.stringify(data),
    })
  },

  async listEntities(params?: {
    limit?: number
    offset?: number
    entity_type?: string
  }): Promise<EntityListResponse> {
    const queryParams = new URLSearchParams()
    if (params?.limit !== undefined) queryParams.append('limit', String(params.limit))
    if (params?.offset !== undefined) queryParams.append('offset', String(params.offset))
    if (params?.entity_type) queryParams.append('entity_type', params.entity_type)
    
    return request<EntityListResponse>(`/knowledge_graph/entities?${queryParams.toString()}`, {
      method: 'GET',
    })
  },

  // 获取实体类型列表
  async listEntityTypes(): Promise<EntityTypesResponse> {
    return request<EntityTypesResponse>('/knowledge_graph/entity-types', {
      method: 'GET',
    })
  },
}

File: rag_backend/app/schemas/knowledge_graph.py
"""
知识图谱相关的 Pydantic 模型
用于 API 请求/响应的数据验证
"""
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field, ConfigDict
from datetime import datetime


# ============ 实体相关 ============
class EntityBase(BaseModel):
    """实体基础模型"""
    name: str = Field(..., description="实体名称")
    type: str = Field(..., description="实体类型: PERSON, LOCATION, ORGANIZATION, CONCEPT")
    properties: Optional[Dict[str, Any]] = Field(default_factory=dict, description="实体属性")


class EntityCreate(EntityBase):
    """创建实体请求"""
    pass


class EntityResponse(EntityBase):
    """实体响应"""
    id: Optional[str] = Field(None, description="Neo4j 节点 ID")
    created_at: Optional[datetime] = None
    
    model_config = ConfigDict(
        from_attributes=True
    )


# ============ 关系相关 ============
class RelationBase(BaseModel):
    """关系基础模型"""
    source: str = Field(..., description="源实体名称")
    target: str = Field(..., description="目标实体名称")
    type: str = Field(..., description="关系类型")
    properties: Optional[Dict[str, Any]] = Field(default_factory=dict, description="关系属性")


class RelationCreate(RelationBase):
    """创建关系请求"""
    pass


class RelationResponse(RelationBase):
    """关系响应"""
    id: Optional[str] = Field(None, description="Neo4j 关系 ID")
    created_at: Optional[datetime] = None
    
    model_config = ConfigDict(
        from_attributes=True
    )


# ============ 图构建相关 ============
class EntityUpdate(BaseModel):
    """Update an existing entity by graph node id."""
    name: str = Field(..., min_length=1, description="实体名称")
    type: str = Field(..., min_length=1, description="实体类型")
    properties: Dict[str, Any] = Field(default_factory=dict, description="实体属性")


class RelationUpdate(BaseModel):
    """Update an existing relation by graph relationship id."""
    source: Optional[str] = Field(None, description="源实体名称")
    target: Optional[str] = Field(None, description="目标实体名称")
    type: str = Field(..., min_length=1, description="关系类型")
    properties: Dict[str, Any] = Field(default_factory=dict, description="关系属性")


class GraphBuildRequest(BaseModel):
    """图构建请求"""
    text: str = Field(..., description="待处理文本")
    user_id: Optional[int] = Field(None, description="用户 ID")
    session_id: Optional[str] = Field(None, description="会话 ID")
    extract_entities: bool = Field(True, description="是否提取实体")
    extract_relations: bool = Field(True, description="是否提取关系")


class GraphBuildResponse(BaseModel):
    """图构建响应"""
    entities: List[EntityResponse] = Field(default_factory=list)
    relations: List[RelationResponse] = Field(default_factory=list)
    success: bool = True
    message: Optional[str] = None


# ============ 图查询相关 ============
class EntityQueryRequest(BaseModel):
    """实体查询请求"""
    entity_name: str = Field(..., description="实体名称")
    max_depth: int = Field(2, ge=1, le=5, description="最大查询深度")
    limit: int = Field(10, ge=1, le=100, description="返回结果数量限制")


class RelatedEntity(BaseModel):
    """相关实体"""
    name: str
    type: str
    distance: int = Field(..., description="距离（跳数）")
    relation_path: Optional[List[str]] = Field(None, description="关系路径")


class EntityQueryResponse(BaseModel):
    """实体查询响应"""
    entity: EntityResponse
    related_entities: List[RelatedEntity] = Field(default_factory=list)
    total_count: int = 0


# ============ 混合检索相关 ============
class HybridSearchRequest(BaseModel):
    """混合检索请求"""
    query: str = Field(..., description="查询文本")
    user_id: Optional[int] = None
    session_id: Optional[str] = None
    top_k: int = Field(5, ge=1, le=20, description="返回结果数量")
    vector_weight: float = Field(0.7, ge=0, le=1, description="向量检索权重")
    graph_weight: float = Field(0.3, ge=0, le=1, description="图检索权重")
    use_graph: bool = Field(True, description="是否使用图检索")


class SearchResult(BaseModel):
    """检索结果"""
    content: str
    score: float
    source: str = Field(..., description="来源: vector, graph, hybrid")
    metadata: Optional[Dict[str, Any]] = Field(default_factory=dict)


class HybridSearchResponse(BaseModel):
    """混合检索响应"""
    results: List[SearchResult] = Field(default_factory=list)
    vector_results_count: int = 0
    graph_results_count: int = 0
    total_count: int = 0


# ============ 图统计相关 ============
class GraphStatsResponse(BaseModel):
    """图统计响应"""
    total_entities: int = 0
    total_relations: int = 0
    entity_types: Dict[str, int] = Field(default_factory=dict)
    relation_types: Dict[str, int] = Field(default_factory=dict)


# ============ 图可视化相关 ============
class GraphNode(BaseModel):
    """图节点（用于可视化）"""
    id: str
    label: str
    type: str
    properties: Dict[str, Any] = Field(default_factory=dict)


class GraphEdge(BaseModel):
    """图边（用于可视化）"""
    id: str
    source: str
    target: str
    type: str
    properties: Dict[str, Any] = Field(default_factory=dict)
    description: Optional[str] = Field(None, description="关系语义描述")


class GraphVisualizationResponse(BaseModel):
    """图可视化响应"""
    nodes: List[GraphNode] = Field(default_factory=list)
    edges: List[GraphEdge] = Field(default_factory=list)
    center_node: Optional[str] = None


class GraphImportRequest(BaseModel):
    """Persist an edited graph snapshot without removing unrelated graph data."""
    nodes: List[GraphNode] = Field(default_factory=list, description="要保存的节点")
    edges: List[GraphEdge] = Field(default_factory=list, description="要保存的关系")
    deleted_node_ids: List[str] = Field(default_factory=list, description="要删除的节点 ID")
    deleted_edge_ids: List[str] = Field(default_factory=list, description="要删除的关系 ID")


class GraphImportResponse(BaseModel):
    """Graph editor save/import result."""
    success: bool = True
    nodes_saved: int = 0
    edges_saved: int = 0
    nodes_deleted: int = 0
    edges_deleted: int = 0
    errors: List[str] = Field(default_factory=list)


class EntityListItem(BaseModel):
    """实体列表项"""
    id: str = Field(..., description="Neo4j 节点 ID")
    name: str = Field(..., description="实体名称")
    type: str = Field(..., description="实体类型")
    properties: Dict[str, Any] = Field(default_factory=dict, description="实体属性")
    created_at: Optional[str] = Field(None, description="创建时间")
    updated_at: Optional[str] = Field(None, description="更新时间")


class EntityListResponse(BaseModel):
    """实体列表响应"""
    entities: List[EntityListItem] = Field(default_factory=list)
    total: int = Field(0, description="总数")
    limit: int = Field(200, description="每页数量")
    offset: int = Field(0, description="偏移量")


class EntityTypesResponse(BaseModel):
    """实体类型列表响应"""
    types: List[str] = Field(default_factory=list)


# ============ 路径查询相关 ============
class PathRequest(BaseModel):
    """路径查询请求"""
    source: str = Field(..., min_length=1, description="源实体名称")
    target: str = Field(..., min_length=1, description="目标实体名称")
    max_depth: int = Field(4, ge=1, le=6, description="最大路径深度")


class PathEntity(BaseModel):
    """路径中的实体"""
    name: str
    type: str


class PathResult(BaseModel):
    """路径查询结果（一条路径）"""
    entities: List[PathEntity] = Field(..., description="路径上的实体列表（有序）")
    relations: List[str] = Field(..., description="路径上的关系类型列表（有序）")
    hops: int = Field(..., description="跳数")


class PathResponse(BaseModel):
    """路径查询响应"""
    source: str
    target: str
    paths: List[PathResult] = Field(default_factory=list)
    total_paths: int = 0
